triedb/pathdb/nodes.go

```diff
@@ -424,3 +427,240 @@ func (s *nodeSetWithOrigin) decode(r *rlp.Stream) error {
 	s.computeSize()
 	return nil
 }
+
+// encodeNodeCompressed encodes the trie node differences between two consecutive
+// versions into byte stream. The format is as below:
+//
+// - metadata byte layout (1 byte):
+//
+//	┌──── Bits (from MSB to LSB) ───┐
+//	│ 7 │ 6 │ 5 │ 4 │ 3 │ 2 │ 1 │ 0 │
+//	└───────────────────────────────┘
+//	  │   │   │   │   │   │   │   └─ FlagA: marks if value is encoded in compressed format (1 always)
+//	  │   │   │   │   │   │   └───── FlagB: marks if no extended bitmap used after the metadata byte
+//	  │   │   │   │   │   └───────── FlagC: bitmap for node (only used when flagB == 1)
+//	  │   │   │   │   └───────────── FlagD: bitmap for node (only used when flagB == 1)
+//	  │   │   │   └───────────────── FlagE: reserved
+//	  │   │   └───────────────────── FlagF: reserved
+//	  │   └───────────────────────── FlagG: reserved
+//	  └───────────────────────────── FlagH: reserved
+//
+// Example:
+//
+// 0b_0000_1011
+//
+// Bit0=1, Bit1=1 -> node in compressed format, no extended bitmap
+// Bit2=0, Bit3=1 -> the key of a short node is not stored; its value is stored.
+//
+// - 2 bytes extended bitmap (only if the flagB in metadata is 0), each bit
+// represents a corresponding child;
+//
+// - concatenation of original value of modified children along with its size;
+func encodeNodeCompressed(addExtension bool, elements [][]byte, indices []int) []byte {
+	var (
+		enc  []byte
+		flag = byte(1) // flagA
+	)
+	if !addExtension {
+		flag |= 2 // flagB
+
+		// Embedded bitmap
+		for _, pos := range indices {
+			flag |= 1 << (pos + 2) // flagC and flagD
+		}
+		enc = append(enc, flag)
+	} else {
+		// Extended bitmap
+		bitmap := make([]byte, 2) // bitmaps for at most 16 children
+		for _, pos := range indices {
+			// Children[16] is only theoretically possible in the Merkle-Patricia-trie,
+			// in practice this field is never used in the Ethereum case.
+			if pos == 16 {
```

Comment:
```suggestion
			if pos >= 16 {
```

Reply:
My fuzzer is hitting this pretty frequently, I would rather panic here, then run into the index out ouf bounds in line 482